### PR TITLE
[GPR] Describe 3D and Simplification properties of mesh layer

### LINKED ISSUE #4806: Request in QGIS ([Feature][Mesh][3D] new 3D mesh layer renderer)
### Request for documentation
From pull request QGIS/qgis#33837
Author: @vcloarec
QGIS version: 3.12

**[Feature][Mesh][3D] new 3D mesh layer renderer**

### PR Description:
This PR change the 3D mesh renderer approach
The new approach is based on OpenGL shader. By this way, the GPU handles rendering wireframe and color ramp shading.

canyoning

There are two entries to render the mesh layer in the 3D view :
- choose the mesh as terrain in the config widget
- activate the 3D view in the layer properties.

The user can choose those settings
- enable/disable smooth triangles
- enable/disable wireframe
- choose the line width and the line color of the wireframe
- change the verticale scale
- choose the style of the rendering : unique color or color ramp shading
- settings the color ramp shading as for raster layer

For now, the mesh can be rendered as a TIN but rendering dataset is not supported. However the infrastructure is adapted for a future PR.
mesh_3d_settings

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [X] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### Commits tagged with [need-docs] or [FEATURE]

### LINKED ISSUE #4937: Request in QGIS ([FEATURE] mesh simplification)
### Request for documentation
From pull request QGIS/qgis#34532
Author: @vcloarec
QGIS version: 3.14

**[FEATURE] mesh simplification**

### PR Description:
As mesh layer could have millions of triangles, rendering could be very slow, especially when all the triangles are displayed in the view whereas triangles are too small to be viewed. 

For those situations, this PR provides an option to simplify the mesh. Simplification leads to one or more simplified mesh that represents levels of detail. When rendering the mesh, the appropriate level of detail is chosen to have an adequate rendering depending on the view.

A new tab in the mesh layer properties widget allows the user to change the settings :
- The reduction factor is used to simplify mesh, each level of details has approximately a number of triangles equal the number of triangles of the previous level of details divided by the reduction factor
- Maximum mesh resolution: this is the average size (in pixels) of the triangles that is permitted to display, if this average size of the mesh is lesser than this value, a lower level of details mesh is displayed, which will have a average size just greater than the maximum mesh resolution.

meshSimplificationWidget

This feature permits to speed up rendering as in those tables (result in milliseconds) :
meshSimplicationBenchmark

Demo GIF
Before:
meshSimplificationbefore

After:
meshSimplificationAfter

### Commits tagged with [need-docs] or [FEATURE]

### LINKED ISSUE #4941: Request in QGIS ([FEATURE] Mesh dataset in 3D view)
### Request for documentation
From pull request QGIS/qgis#34675
Author: @vcloarec
QGIS version: 

**[FEATURE] Mesh dataset in 3D view**

### PR Description:
Time being, mesh layer can be rendered in 3D view only based on the Z values of the mesh's vertices to render the terrain.
This PR adds the possibility to render a mesh layer in the 3D view based on values of the dataset. 
The vertical component of the vertices can be set equal to dataset values (for example, level of water surface) and the texture of the mesh can be set to render other dataset values with color shading (for example velocity).
The entry to render dataset in 3D view is the layer properties of the mesh layer. 

image

So, the principle to render the two types of rendering (Z value and dataset) is to use two entries ;
- the terrain settings are in the 3D configuration widget, as before this PR
- the dataset rendering settings are in the mesh layer properties, the user can choose:
                      - the dataset group that will be used for the vertical component of the mesh
                      - the vertical scale
                      - If the dataset has to be considered as relative to the vertices Z value
                      - The rendering color settings : unique color, or with the color ramp shader already used for 2D shading

That permits to render the terrain and the dataset of a mesh layer in the same 3D view.

image

### Commits tagged with [need-docs] or [FEATURE]

### LINKED ISSUE #4998: Request in QGIS ([FEATURE][3D][MESH] arrows for 3D mesh layer dataset rendering)
### Request for documentation
From pull request QGIS/qgis#34945
Author: @vcloarec
QGIS version: 3.14

**[FEATURE][3D][MESH] arrows for 3D mesh layer dataset rendering**

### PR Description:
This PR adds the possibility to display arrows on mesh layer dataset 3D entity, depending on vector dataset.
The dataset used to render arrows is the one chosen for rendering vectors on 2D rendering.
In the same way, the color of arrows are defined in the 2D rendering settings for vector dataset.
In 3D settings, the user can define the spacing of arrows and if size is fixed or scaled on magnitude. 
This spacing setting defines also the max size of arrows because arrows can't overlap.
arrows_fixed_size
arrows_scaled_size

### Commits tagged with [need-docs] or [FEATURE]

### PR DESCRIPTION
**Pull request by DelazJ**
Should fix #4941 fix #4806 fix #4937 fix #4998
Again some features I couldn't understand their function

### CODE REVIEW

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 1]
```diff
@@ -302,6 +302,79 @@ as shown in :numref:`figure_mesh_symbology_grid`:
...
+   :align: center
+
+   Mesh dataset 3D properties
+
+Check |checkbox| :guilabel:`Enable 3D Renderer` and you can edit following
+properties:
+
+* Under :guilabel:`Triangle settings`
+
+  * :guilabel:`Smooth triangles`: ???
```
Comment (PeterPetrik) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:327]: @vcloarec what are smooth triangles?

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 2]
```diff
@@ -302,6 +302,79 @@ as shown in :numref:`figure_mesh_symbology_grid`:
...
+  * :guilabel:`Show wireframe` whose you can set the :guilabel:`Line width`
+    and :guilabel:`Color`
+  * :guilabel:`Level of detail`: ???
```
Comment (PeterPetrik) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:330]: in the layer properties -> rendering -> simplify mesh you can create something like raster overviews for mesh. So when you zoom out, some triangles are simplified and "merged" to bigger chunks. this helps a lot for 2d/3d rendering. Here you can choose which "level of detail" you want to use from the set.

File: docs/user_manual/working_with_mesh/mesh_properties.rst (hunk 1 above)
Comment (vcloarec) [reply docs/user_manual/working_with_mesh/mesh_properties.rst:327]: visual effect that smooths the angles between triangles.

Not smoothed:
image

Smoothed:
image

File: docs/user_manual/working_with_mesh/mesh_properties.rst (hunk 2 above)
Comment (DelazJ) [reply docs/user_manual/working_with_mesh/mesh_properties.rst]: opened a report for this at https://github.com/qgis/QGIS/issues/44868

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 3]
```diff
@@ -423,6 +423,94 @@ is shown in the dialog but you can read more on the methods at
...
+    smoothed for a better 3D rendering
+  * :guilabel:`Show wireframe` whose you can set the :guilabel:`Line width`
+    and :guilabel:`Color`
+
+  .. _levelofdetail:
+
+  * :guilabel:`Level of detail`: Controls how :ref:`simplified <meshrendering>`
+    the mesh layer to render should be. On the far right, it is the base mesh,
+    and the more you go left, the more the layer is simplified and is rendered
+    with less details.
```
Comment (vcloarec) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:461]: Add:
    This option is only available if 'Simplify mesh' option is activated.

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 4]
```diff
@@ -423,6 +423,94 @@ is shown in the dialog but you can read more on the methods at
...
+* :guilabel:`Vertical settings` to control what to display
```
Comment (vcloarec) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:462]:
```suggestion
* :guilabel:`Vertical settings` to control behavior of the vertical component of vertices of rendered triangles
```

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 5]
```diff
@@ -423,6 +423,94 @@ is shown in the dialog but you can read more on the methods at
...
+
+  * :guilabel:`Dataset group for vertical value`: the dataset group that will
+    be used for the vertical component of the mesh
+  * |unchecked|:guilabel:`Dataset value relative to vertices Z value`: whether
+    to consider the dataset values as absolute Z coordinate or relative to
+    the vertices Z value
```
Comment (vcloarec) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:468]:
```suggestion
    the vertices native Z value
```

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 6]
```diff
@@ -423,6 +423,94 @@ is shown in the dialog but you can read more on the methods at
...
+  also the max size of arrows because arrows can't overlap.
+
+
+.. index:: Rendering
+.. _meshrendering:
+
+Rendering Properties
+--------------------
+
+As mesh layers can have millions of triangles, their rendering can sometimes
```
Comment (vcloarec) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:489]:
```suggestion
As mesh layers can have millions of faces, their rendering can sometimes
```

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 7]
```diff
@@ -423,6 +423,94 @@ is shown in the dialog but you can read more on the methods at
...
+be very slow, especially when all the triangles are displayed in the view
```
Comment (vcloarec) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:490]:
```suggestion
be very slow, especially when all the faces are displayed in the view
```

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 8]
```diff
@@ -423,6 +423,94 @@ is shown in the dialog but you can read more on the methods at
...
+whereas they are too small to be viewed.
+To speed up the rendering, you can simplify the mesh layer, resulting in one
+or more meshes representing different :ref:`levels of detail <levelofdetail>`
+and select at which level of detail you would like QGIS to render the mesh layer.
```
Comment (vcloarec) [inline docs/user_manual/working_with_mesh/mesh_properties.rst:494]:
```suggestion
and select at which level of detail you would like QGIS to render the mesh layer.
Note that the simplify mesh contains only triangular faces.
```

File: docs/user_manual/working_with_mesh/mesh_properties.rst [hunk 9]
```diff
@@ -302,6 +302,89 @@ as shown in :numref:`figure_mesh_symbology_grid`:
...
+  the second 50 000 faces, the third 5000,...
+  If a higher reduction factor leads quickly to simpler meshes (i.e. with triangles
+  of bigger size), it produces also fewer levels of detail.
+* :guilabel:`Maximal mesh resolution`: the average size (in pixels)
+  of the triangles that is permitted to display. If the average size of the
+  mesh is lesser than this value, the rendering of a lower level of details
+  mesh is triggered.
+
+  .. I tried to reformulate simply https://github.com/qgis/QGIS-Documentation/issues/4937.
+   Hope that I caught the idea.
```
[the remark this answers is not in the document]
Comment (vcloarec) [reply docs/user_manual/working_with_mesh/mesh_properties.rst:512]: yes, you are right, this is confusing. The label is wrong, this should be "Minimum mesh resolution". 

And in hindsight, I am wondering if we set the label with "Minimum triangle size" instead will not be clearer. In this case, we need to change also the suffix in the gui from "pixels/triangle" to "pixels"

**Review by vcloarec — CHANGES_REQUESTED** [review]

File: docs/user_manual/working_with_mesh/mesh_properties.rst (hunk 3 above)
Comment (DelazJ) [reply docs/user_manual/working_with_mesh/mesh_properties.rst:461]:
```suggestion
    with less details.
    This option is only available if the :guilabel:`Simplify mesh` option under
    the :guilabel:`Rendering` tab is activated.
```

File: docs/user_manual/working_with_mesh/mesh_properties.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (DelazJ) [reply docs/user_manual/working_with_mesh/mesh_properties.rst:512]:
```suggestion
* :guilabel:`Minimum triangle size`: the average size (in pixels)
  of the triangles that is permitted to display. If the average size of the
  mesh is lesser than this value, the rendering of a lower level of details
  mesh is triggered.

```

**Comment by DelazJ** [comment]
Thanks a lot @PeterPetrik and @vcloarec for the reviews.

**Pull request merged**